### PR TITLE
Add support for inferring component URLs in the global defaults repository

### PR DESCRIPTION
**Pull request by simu**
This PR adds support to infer component URLs by running `commodore inventory components` for component version pins which don't have a component URL themselves.

### CODE REVIEW

File: src/commodore/inventory.spec.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      const f = new Facts();
+      f.distribution = 'dist';
+      f.cloud = 'cloud';
+      f.region = 'region';
+      const factsPath = await inv.writeFactsFile('ck', f);
+
+      expect(factsPath).toBe(`${cacheDir()}/ck-facts.yaml`);
+
+      const resStr = (await readFile(factsPath)).toString();
+      const res = yaml.load(resStr);
```
Comment (glrf) [inline src/commodore/inventory.spec.ts:27]:
```suggestion
      const res: any = yaml.load(resStr) as any;
```
Again my LSP will complain, as `res` has type `unknown` and you technically cannot access any fields of it.

File: src/commodore/index.spec.ts [hunk 2]
```diff
@@ -1,18 +1,49 @@
-import { loadFixture } from '../test/util';
-import { extractPackageFile } from './index';
+import { mkdir } from 'fs/promises';
+import { getFixturePath, loadFixture } from '../test/util';
+import { extractPackageFile, defaultConfig } from './index';
+import { clearCache } from './inventory';
 import { expect, describe, it } from '@jest/globals';
```
Comment (glrf) [inline src/commodore/index.spec.ts:5]:
```suggestion
import { jest, beforeEach, beforeAll, expect, describe, it } from '@jest/globals';
```
You actually use these. Not sure how this works without the imports.

File: src/commodore/index.spec.ts [hunk 3]
```diff
@@ -1,18 +1,49 @@
...
 
 const params1 = loadFixture('1/params.yml');
 const kube2 = loadFixture('2/kubernetes.yml');
 const invalid3 = loadFixture('3/params.yml');
+const pin4 = loadFixture('4/pins.yml');
 
-describe('manager/commodore/index', () => {
+jest.mock('renovate/dist/config/global');
+const renovate_global = require('renovate/dist/config/global');
```
Comment (glrf) [inline src/commodore/index.spec.ts:13]: This i just an import. I would probably use

```
import { getGlobalConfig } from 'renovate/dist/config/global';
```
or
```
import * as renovateGlobal from 'renovate/dist/config/global';
```

File: src/commodore/index.spec.ts [hunk 4]
```diff
@@ -1,18 +1,49 @@
...
+function mockGetGlobalConfig(fixtureId: string) {
+  renovate_global.getGlobalConfig.mockImplementation(() => {
```
Comment (glrf) [inline src/commodore/index.spec.ts:15]: I get a type error. [This stackoverflow post](https://stackoverflow.com/questions/51495473/typescript-and-jest-avoiding-type-errors-on-mocked-functions) shows a few ways to fix this.

**Review by glrf — COMMENTED** [review]
In general this looks good. I did not see any actual bugs.

I'm mainly forwarding input from my LSP.

I think the test files do not actually need to be valid typescript to run with jest. And it seems like we also do not need to import any jest dependencies for it to run. That's why I have so many changes for the tests. We do not really need to fix these issues, but it would still be nice if they were valid typescript, even if its only to prevent my LSP from screaming at me.

File: src/index.ts [hunk 5]
```diff
@@ -2,4 +2,16 @@ import api from 'renovate/dist/manager/api.js';
 import * as commodore from './commodore';
 
 api.set('commodore', commodore);
+
+// Patch renovate option validation to accept `commodore.extraConfig`
+// parameter.
+const options_1 = require('renovate/dist/config/options');
```
[the remark this answers is not in the document]
Comment (simu) [reply src/index.ts]: I thought it didn't work as I wanted it to with the direct import, but I must have made another mistake initially.

File: src/commodore/util.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  }
+  if (facts.region !== null && ignoreValues.includes(facts.region)) {
+    facts.region = null;
+  }
+
+  return facts;
+}
+
+/* Check whether `facts` has any non-null field */
+export function hasFact(facts: Facts): boolean {
```
[the remark this answers is not in the document]
Comment (simu) [reply src/commodore/util.ts]: Done, this was used in the implementation previously, but I removed that part again.

File: src/commodore/inventory.spec.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import yaml from 'js-yaml';
+import { mkdir, readFile, unlink } from 'fs/promises';
+
+import { loadFixture } from '../test/util';
+import { expect, describe, it } from '@jest/globals';
```
[the remark this answers is not in the document]
Comment (simu) [reply src/commodore/inventory.spec.ts:4]: It appears to, since before the `beforeAll(mkdir)` the tests didn't work on GitHub.

**Comment by simu** [comment]
@glrf can you please check again? I think I fixed everything now, but  I don't fully trust my LSP setup yet.

**Review by glrf — APPROVED** [review]
LGTM

File: src/commodore/inventory.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  let cachedVersions = versionCache.get(ck);
+  if (cachedVersions !== undefined) {
+    logger.info(`Reusing cached versions for ${ck}`);
+    return cachedVersions;
+  }
+
+  const factsPath: string = await writeFactsFile(ck, facts);
+
+  /* construct and execute Commodore command */
+  var command = `commodore inventory components ${repoPath} -ojson -f ${factsPath} -f ${extraValuesPath}`;
```
Comment (bastjan) [inline src/commodore/inventory.ts:50]: I'm offended that it's the year 2021 and the abysmal node standard library still does not support passing arguments as an array.

File: src/commodore/index.ts [hunk 9]
```diff
@@ -2,30 +2,43 @@ import yaml from 'js-yaml';
...
+export const defaultExtraConfig = {
+  extraParameters: {},
+  distributionRegex:
+    /^distribution\/(?<distribution>[^\/]+)(?:\/cloud\/(?<cloud>.+)\.ya?ml|(?:\/.+)?\.ya?ml)$/,
+  cloudRegionRegex:
+    /^cloud\/(?<cloud>[^\/]+)(?:\/(?<region>.+)\.ya?ml|\.ya?ml)$/,
+  ignoreValues: ['params'],
+};
 
 // exctractComponents will extract all component dependencies.
```
Comment (bastjan) [inline src/commodore/index.ts:34]:
```suggestion
// extractComponents will extract all component dependencies.
```

**Review by bastjan — APPROVED** [review]
Looks good! 🚀

File: src/commodore/inventory.ts (hunk 8 above)
Comment (simu) [reply src/commodore/inventory.ts:50]: It does for `spawn()` (cf. https://nodejs.org/api/child_process.html#child_processspawncommand-args-options), but not for `exec()`.

**Pull request merged**